File: tools/python/ir/nodes.py
"""STUNIR Semantic IR Node Structures.

DO-178C Level A Compliant
Base node types and type references.
"""

from enum import Enum
from typing import Optional, Dict, Any
from pydantic import BaseModel, Field

from .ir_types import NodeID, IRHash, IRName, SourceLocation, IRNodeKind, IRPrimitiveType


class TypeKind(str, Enum):
    """Type kind discriminator."""
    PRIMITIVE = "primitive_type"
    ARRAY = "array_type"
    POINTER = "pointer_type"
    STRUCT = "struct_type"
    FUNCTION = "function_type"
    REF = "type_ref"


class TypeReference(BaseModel):
    """Base type reference."""
    kind: TypeKind

    class Config:
        use_enum_values = True


class PrimitiveTypeRef(TypeReference):
    """Primitive type reference."""
    kind: TypeKind = TypeKind.PRIMITIVE
    primitive: IRPrimitiveType


class TypeRef(TypeReference):
    """Named type reference."""
    kind: TypeKind = TypeKind.REF
    name: IRName
    binding: Optional[NodeID] = None


class IRNodeBase(BaseModel):
    """Base IR node structure."""
    node_id: NodeID
    kind: IRNodeKind
    location: Optional[SourceLocation] = None
    type: Optional[TypeReference] = None
    attributes: Dict[str, Any] = Field(default_factory=dict)
    hash: Optional[IRHash] = None

    class Config:
        use_enum_values = True
        arbitrary_types_allowed = True

File: tools/rust/src/nodes.rs
//! STUNIR Semantic IR Node Structures
//!
//! Base node types and type references.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::types::*;

/// Type kind discriminator
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TypeKind {
    PrimitiveType,
    ArrayType,
    PointerType,
    StructType,
    FunctionType,
    TypeRef,
}

/// Type reference
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case")]
pub enum TypeReference {
    PrimitiveType {
        primitive: IRPrimitiveType,
    },
    TypeRef {
        name: IRName,
        #[serde(skip_serializing_if = "Option::is_none")]
        binding: Option<NodeID>,
    },
    // Simplified for now - can be extended
}

/// Base IR node structure
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IRNodeBase {
    pub node_id: NodeID,
    pub kind: IRNodeKind,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub location: Option<SourceLocation>,
    #[serde(skip_serializing_if = "Option::is_none", rename = "type")]
    pub type_info: Option<TypeReference>,
    #[serde(default, skip_serializing_if = "HashMap::is_empty")]
    pub attributes: HashMap<String, serde_json::Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hash: Option<IRHash>,
}

impl IRNodeBase {
    /// Create a new IR node base
    pub fn new(node_id: NodeID, kind: IRNodeKind) -> Self {
        Self {
            node_id,
            kind,
            location: None,
            type_info: None,
            attributes: HashMap::new(),
            hash: None,
        }
    }

    /// Check if node ID is valid
    pub fn is_valid_node_id(node_id: &str) -> bool {
        node_id.starts_with("n_") && node_id.len() > 2
    }

    /// Check if hash is valid
    pub fn is_valid_hash(hash: &str) -> bool {
        hash.starts_with("sha256:") && hash.len() == 71
    }
}
